### ISSUE TITLE
PostgreSQL : First-class TimescaleDB support

### ISSUE THREAD

**Issue by Hebilicious** [prisma/prisma#3228]
## Problem
TimescaleDB is an incredible postgreSQL extension that allows you to work with time-series data in native SQL. It's amazing for IOT heavy projects and there's plenty of other use cases.
It's possible to use TimescaleDB and prisma together with some manual work and using raw queries.

- This was already possible with prisma1 as stated here : https://github.com/prisma/prisma1/issues/1672#issuecomment-511410796

- This has been made easier with the implementation of composite keys : https://github.com/prisma/prisma/issues/1389 🎉

However it is still needed to manually install the extension, then manually convert tables into hypertables using raw sql, and some TimescaleDB features are implemented using [Postgres Functions](https://www.postgresql.org/docs/current/sql-createfunction.html).

I haven't tested it personally, but going by the docs prisma should work out of the box once the hypertable is properly created when it comes to writing data.
For reading data, falling back to raw sql is needed for some of the features implemented with functions (docs [here](https://docs.timescale.com/latest/using-timescaledb/reading-data)).

## Suggested solution

It would be nice if the prisma schema supported the following :

- Ability to add the extension from the prisma schema directly
- Ability to define a hypertable directly, so that prisma migrate would handle them without having to fallback to raw sql after running a migration

Regarding the prisma client, it would be great if it supported the use of functions.
I don't know what the syntax would look like, but in raw sql it's pretty straightforward: 

```sql
SELECT location, last(temperature, time)
  FROM conditions
  GROUP BY location;
```
Function support might need it's own issue thread, but I'm curious to see what a potential client API would look like.
Not that in this case, prisma doesn't support groupBy yet (https://github.com/prisma/prisma-client-js/issues/1)

It might be worth mentioning than the schema could also support function creation as well, however this would be out of the scope of the TimescaleDB support, as the function are created when the extension is installed.

## Alternatives

It's still possible to use raw sql and manually handle the hypertable creation/functions call 

## Additional context

It would be really nice to have a more generic approach to add some of the most popular extensions directly though the prisma schema, postGIS comes to mind. Right now if you want to use prisma + postgis + timescale together, there's a lot of manual wiring to do.

Relevant : 

- Postgis support : https://github.com/prisma/prisma/issues/2789
- Native DataTypes : https://github.com/prisma/prisma/issues/446

I do not know if something needs to happen on the timescaleDB end, as it is just a regular postgreSQL extension (as far as I know). However @akulkarni from the timescaleDB team said he was willing to help.

**Comment by Sytten** [comment]
I also love timescaledb and feel the same pain points. I think that is out of scope for the prisma "core". What I think would be awesome is if the prisma team gears the project more toward a plugin architecture similar to what Hashicorp did with terraform providers. So the community could implement support for new databases, new primitives, etc.

**Comment by Hebilicious** [comment]
Since timescaledb/postgis are native postgres extensions, I believe they should be in prisma scope at some point.
In the native datatype thread, the proposed syntax looks like this : 
```prisma
datasource mydb {
  provider = "postgres"
  url      = "..."
}

model Blog {
  id       Int      @id
  createAt DateTime @mydb.Timestamp
}
```

maybe we could have something like 

```prisma
datasource mydb {
 provider = "postgres"
extensions = ["timescaleDB"]
...
}

model Conditions {
time DateTime @mydb.type("TIMESTAMPTZ") //native postgres type
location String @mydb.type("TEXT") //native postgres type
@@mydb.extensions["timescaleDB"].hypertable("time") //AFAIK that's the only needed thing, and that maps to SELECT create_hypertable('conditions', 'time');
}
```

Similarly for postGIS, there's a large  list of [functions](https://postgis.net/docs/reference.html#Measurement_Functions), so having an extension agnostic way to deal with postgres functions from the prisma client would be a great feature to have.

**Comment by akulkarni** [comment]
👋🏽 Hi there, Ajay from TimescaleDB. If there are things we can to do help with this issue please let us know. We also have an active community on Slack: https://slack.timescale.com/

**Comment by vorillaz** [comment]
I am a little late to the party but I am trying to build an application using Prisma along with Timescale. The thing is that I am a little confused regarding using the schema and Prisma migrations. As far as I can tell I can't use the schema declaration defining the schema changes as Prisma does now allow declaring a model without using `@id` or `@unique`. 
I am basically trying to follow the [Node.js + Timescale tutorial](https://docs.timescale.com/latest/tutorials/quickstart-node).

I love the suggested solution and I would love to have some feedback or a sample repository if anyone's already tried using Prisma along with Timescale cc @akulkarni

**Comment by janpio** [comment]
(🌃 🐰 Although I work for Prisma during the week and day, I write this as a private person that played around with Prisma and TimescaleDB during the Easter brake - so please do not consider this _official_ information.)

As TimescaleDB is really just PostgreSQL _plus_ an extension, I was quite positive this would _just work_ ™️ and played with it a bit to learn how TimescaleDB works and what I can do with it.

While doing so, I used the above linked "Node + Timescale" tutorial. As I of course adapted that for Prisma, I created a small Gist based off the original that shows the current steps necessary for Prisma:
https://gist.github.com/janpio/2a425f22673f2de54469772f16af8118

The diff between the original and the Prisma version gives a good overview that not much really had to be changed, but this is really just doing it "the Prisma way", using the appropriate CLI commands and Prisma Client code: https://gist.github.com/janpio/2a425f22673f2de54469772f16af8118/revisions#diff-15f6c3367f1237852c83d79bf8e0fd698a694a378ded36abf802fbef4b024cc2

Conclusion: With the new Prisma Migrate, activating the extension and creating a Hypertable is pretty simple - it executes a plain SQL file after all. Prisma Client works as expected against that hypertable.

I did not get to adding a TimescaleDB native function call via a raw query yet.

**Comment by vorillaz** [comment]
Hey, @janpio thanks for the lovely comment.
My main concern with the current state of Prisma along with Timescale is that due to Prisma schema limitation we are creating inconsistencies between the models and the actual DB schema.
More precisely since models without `ids` are not supported. They should be dropped through the migration script or even worse models should not include the timescale tables.

**Comment by janpio** [comment]
Ah, I wanted to reply to that: 
Tables/models without a primary key or Id are supported, you just need _any_ unique identifier for the data. In my version of the tutorial, I chose a compound unique key across both fields. You usually always want to have a way to uniquely identify data in a table, so this makes sense to add in my opinion anyway.

**Comment by vorillaz** [comment]
Yeah, that makes sense, but I guess depends on the use case.
In my scenario, I am using the hyper for analytics parsing thus there is no need for ids or unique keys, just plain and fast writes. Following the docs from Prisma v1 I am dropping the ids when migrating but that does not feel right IMHO.

**Comment by janpio** [comment]
> Following the docs from Prisma v1 I am dropping the ids when migrating but that does not feel right IMHO.

Whatever is documented for Prisma 1 does not apply to Prisma 2.

**Comment by Hebilicious** [comment]
@janpio Just chiming in, I'm not working on any IOT related project or anything, but if I'm not mistaken in the use case where you are collecting metrics coming from a lot of sensors, you just might not want to have any kind of unique identifier on them.

Maybe @akulkarni can give us more information regarding whether not having a unique index on your hypertable is something desirable? 

Out of curiosity, would it be complicated for prisma to implement a way to query models without a unique identifier ?

**Comment by janpio** [comment]
> Out of curiosity, would it be complicated for prisma to implement a way to query models without a unique identifier ?

Yes, that is unfotunately one of the fundamental limitations of Prisma: https://www.prisma.io/docs/about/limitations#records-must-be-uniquely-identifiable Of course this has been a choice, so could be removed - but it is a very deep assumption in all our logic.

Having some feedback from @akulkarni on how to get hypertables, which usually might be fine without any unique identifiers, and this requirement from Prisma together would be nice.

Until then, I would suggest you try adding some unique identifier to that table and see if that could actually be fine for you.

**Comment by pfried** [comment]
@janpio Thank you for your gist, tried to follow along and it makes sense to me. I do have unique indexes so that is fine.

Since timescale is moving towards Timescale Forge (or Cloud) I do now have the issue that Prisma seems to need some permissions which the Timescale Forge user does not have: `Error querying the database: db error: ERROR: must be owner of view pg_stat_statements`.

Is there any overview which permissions /requirements are needed? @akulkarni I think it would be greate to setup some kind of tutorial or blog post around this topic, maybe we can figure something out here?

**Comment by janpio** [comment]
Interesting, I can find no mention of `pg_stat_statements` in our codebase.

**Comment by Sytten** [comment]
If I remember correctly, Timescale does some magic to present a unified table as the frontend of multiple chunks/tables. This requires your user to have more permissions than a regular user usually has for normal tables. I am almost 100% sure this has nothing to do with prisma.

**Comment by pfried** [comment]
Okay, so what happened was that I ran the migration having the `CREATE EXTENSION IF NOT EXISTS timescaledb CASCADE;` statement which failed (already installed error). Since the statement is kind of pointless on timescale forge I dropped it but somehow a migration shift was detected, resulting in a required reset which then resulted in:

    npx prisma migrate dev
    Environment variables loaded from .env
    Prisma schema loaded from prisma\schema.prisma
    Datasource "db": PostgreSQL database "tsdb", schema "public" at "123.123.tsdb.forge.timescale.com:30539"
    
    √ Drift detected: Your database schema is not in sync with your migration history.
    
    We need to reset the PostgreSQL database "tsdb" at "123.123.tsdb.forge.timescale.com:30539".
    Do you want to continue? All data will be lost. ... yes
    
    Error: P3016
    
    The fallback method for database resets failed, meaning Migrate could not clean up the database entirely. Original error: 
    Database error
    Error querying the database: db error: ERROR: must be owner of view pg_stat_statements
       0: sql_migration_connector::best_effort_reset
               with connection=Connection(Postgres(PostgresUrl { url: Url { scheme: "postgres", username: "tsdbadmin", password: Some("XXXXXXX"), host: Some(Domain("123.123.tsdb.forge.timescale.com")), port: Some(30539), path: "/tsdb", query: Some("sslmode=require&statement_cache_size=0"), fr
    agment: None }, query_params: PostgresUrlQueryParams { ssl_params: SslParams { certificate_file: None, identity_file: None, identity_password: <HIDDEN>, ssl_accept_mode: AcceptInvalidCerts }, connection_limit: None, schema: "public", ssl_mode: Require, pg_bouncer: false, host: None, socket_timeout: None, connec
    t_timeout: Some(5s), pool_timeout: Some(10s), statement_cache_size: 0, max_connection_lifetime: None, max_idle_connection_lifetime: Some(300s) } }))
                 at migration-engine\connectors\sql-migration-connector\src\lib.rs:99
       1: migration_core::api::Reset
                 at migration-engine\core\src\api.rs:157

Most likely not really a prisma issue but might be some assumption made about a postgres db. I do not want to make big trouble but I would really like to get a solution based on Timescale + Prisma going, if someone is interested please contact me.

**Comment by pfried** [comment]
I do not want to clutter this issue further so I created a discussion here: https://github.com/prisma/prisma/discussions/6963

@janpio Would be really great if you could have a quick look!

**Comment by pfried** [comment]
Related issue since an application using timescale will most likely use materialized views: https://github.com/prisma/prisma/issues/678 currently the ony sensible workaround is raw sql

I am using prisma with timescale now and it basically works well if you look at it as a postgresql database, the most heavylifiting is done in a custom sql migration, as discussed earlier. As I see it now there are two blocking things:

* Extension issue described above (most likely only true for timescale cloud DaaS offering)
* Views support (If there was just an annotaion avoiding the creation of table, that would basically be enough for now)

**Comment by janpio** [comment]
PostgreSQL extension management we are tracking in https://github.com/prisma/prisma/issues/12308

**Comment by leppaott** [comment]
Yeah a schema without @id/@unique would be nice. The problem is that ID<->TIMESTAMP index doesn't work if we have many items on the same millisecond etc.

**Comment by leppaott** [comment]
@pfried what do you get by using `extensions = [timescaledb]` when you still need migration to do everything (create hypertable)? Does that help with something else?

**Comment by pfried** [comment]
@leppaott This is part of Prismas extension management (e.g. inspect db), see #12308). If you call Timescale function in your migration file without activating the extiensions it will simply error saying that those function (e.g. hypertable stuff) do not exist

**Comment by leppaott** [comment]
@pfried right but still use `CREATE EXTENSION IF NOT EXISTS "timescaledb";` on the migration. But I see now that it's automatically created however since you need to manually edit the file anyway...

**Comment by pfried** [comment]
@leppaott I think the whole point of inspection / pulling is to remove more and more manual steps (and by this possible errors too)

**Comment by Krister-Johansson** [comment]
I ran into this same gap, so I built a small Prisma Client extension to make TimescaleDB nicer to work with: prisma-extension-timescaledb (https://www.npmjs.com/package/prisma-extension-timescaledb).

It's brand-new and community-built (I'm the author), not an official solution, but it covers some of what's asked for in this thread like defining hypertables, continuous aggregates, compression and retention policies without hand-writing all the raw SQL. Would love feedback or bug reports from anyone here who's been doing this manually. 👋
